### PR TITLE
Flow Matching Surface Head: generative pressure prediction (AlphaFold3-inspired)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The entire research programme treats surface pressure prediction as **deterministic regression**: one input → one output. But the hardest metric — p_tan (aft-foil tandem pressure) — may fail because the conditional distribution of aft-foil pressure given (geometry, Re, gap, stagger) has high variance or multiple modes. When forced to predict a single point estimate, the model regresses to the conditional mean, which may be far from any physically realizable solution.

**Flow matching** (Lipman et al., 2022) replaces the deterministic SRF with a generative model that learns the full conditional distribution p(pressure_surface | backbone_hidden). At inference time, we sample N=8 predictions from the flow and average them, producing a statistically consistent estimate more robust to multi-modality than a point estimate.

**This is the approach used in AlphaFold3** — the structure module uses a diffusion head for per-residue coordinates instead of deterministic regression. We apply the exact same principle to surface pressure prediction.

**Critical detail:** The flow matching head operates ONLY on surface nodes (M_s << N_total), keeping computation tractable. It replaces the existing SRF head's loss, not the SRF architecture entirely.

**Target metrics:** p_tan (multi-modal tandem pressure), p_oodc (OOD samples with high prediction variance)

**References:**
- Lipman et al. "Flow Matching for Generative Modeling" (arXiv:2210.02747, 2022)
- AlphaFold3 (Abramson et al., Nature 2024) — diffusion head for structure coordinates
- Tong et al. "Improving and Generalizing Flow-Matching" (arXiv:2302.00482, 2023)

---

### Step 1: Add config flags

```python
flow_matching_surface: bool = False     # enable flow matching for surface prediction
flow_matching_n_samples: int = 8        # inference samples (averaged)
flow_matching_weight: float = 0.5       # weight of flow matching loss vs standard SRF loss
flow_matching_t_embed_dim: int = 16     # time embedding dimension
```

### Step 2: Implement the flow matching surface head

Add this class near the SurfaceRefinementHead:

```python
class FlowMatchingSurfaceHead(nn.Module):
    """
    Conditional flow matching for surface pressure prediction.
    Learns a velocity field v(x_t, t, h) that maps noise → target surface values.
    
    Training: flow matching objective (no ODE simulation needed).
    Inference: solve ODE from noise to prediction using Euler steps.
    """
    def __init__(self, n_hidden=192, out_dim=3, t_embed_dim=16, hidden_dim=256):
        super().__init__()
        self.out_dim = out_dim
        self.t_embed_dim = t_embed_dim
        
        # Time embedding: sinusoidal
        self.t_mlp = nn.Sequential(
            nn.Linear(t_embed_dim, 64),
            nn.SiLU(),
            nn.Linear(64, 64),
        )
        
        # Velocity predictor: [h (n_hidden) + x_t (out_dim) + t_emb (64)] → v (out_dim)
        self.net = nn.Sequential(
            nn.Linear(n_hidden + out_dim + 64, hidden_dim),
            nn.SiLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.SiLU(),
            nn.Linear(hidden_dim, out_dim),
        )
        # Zero-init last layer for safe start
        nn.init.zeros_(self.net[-1].weight)
        nn.init.zeros_(self.net[-1].bias)
    
    def time_embed(self, t):
        """Sinusoidal time embedding. t: [M] in [0,1]"""
        freqs = torch.exp(torch.linspace(0, -4, self.t_embed_dim // 2, device=t.device))
        args = t.unsqueeze(-1) * freqs.unsqueeze(0) * 2 * 3.14159
        return self.t_mlp(torch.cat([args.sin(), args.cos()], dim=-1))  # [M, 64]
    
    def forward(self, h, x_t, t):
        """
        h: [M_s, n_hidden] — backbone hidden states at surface nodes
        x_t: [M_s, out_dim] — noisy surface values at time t
        t: [M_s] — timesteps in [0, 1]
        Returns: v: [M_s, out_dim] — predicted velocity field
        """
        t_emb = self.time_embed(t)  # [M_s, 64]
        inp = torch.cat([h, x_t, t_emb], dim=-1)  # [M_s, n_hidden + out_dim + 64]
        return self.net(inp)
    
    @torch.no_grad()
    def sample(self, h, n_steps=10):
        """
        Generate predictions via Euler ODE integration from noise.
        h: [M_s, n_hidden]
        Returns: x_1: [M_s, out_dim] — predicted surface values
        """
        M_s = h.shape[0]
        x = torch.randn(M_s, self.out_dim, device=h.device)  # start from noise
        dt = 1.0 / n_steps
        for i in range(n_steps):
            t = torch.full((M_s,), i * dt, device=h.device)
            v = self.forward(h, x, t)
            x = x + v * dt
        return x


def flow_matching_loss(head, h_surf, y_surf):
    """
    Conditional flow matching training loss.
    No ODE simulation needed — directly regress the velocity field.
    
    h_surf: [M_s, n_hidden] — backbone hidden at surface nodes
    y_surf: [M_s, out_dim] — ground truth surface (Ux, Uy, p)
    Returns: scalar loss
    """
    M_s = h_surf.shape[0]
    
    # Sample random timesteps
    t = torch.rand(M_s, device=h_surf.device)  # [M_s] in [0, 1]
    
    # Sample noise
    x_0 = torch.randn_like(y_surf)  # [M_s, out_dim]
    
    # Linear interpolation: x_t = (1-t) * x_0 + t * y
    t_expand = t.unsqueeze(-1)  # [M_s, 1]
    x_t = (1 - t_expand) * x_0 + t_expand * y_surf
    
    # Target velocity: v_target = y - x_0 (constant for linear interpolation)
    v_target = y_surf - x_0
    
    # Predicted velocity
    v_pred = head(h_surf, x_t, t)
    
    return F.mse_loss(v_pred, v_target)
```

### Step 3: Add to Transolver model

In `Transolver.__init__`:
```python
if cfg.flow_matching_surface:
    self.flow_surface = FlowMatchingSurfaceHead(
        n_hidden=cfg.n_hidden, out_dim=3,
        t_embed_dim=cfg.flow_matching_t_embed_dim
    )
```

### Step 4: Training — add flow matching loss alongside standard loss

In the training loop, after computing standard surface loss, add:

```python
if cfg.flow_matching_surface:
    # h_surf: backbone hidden states at surface nodes [M_s, n_hidden]
    # y_surf: ground truth at surface nodes [M_s, 3]
    fm_loss = flow_matching_loss(model.flow_surface, h_surf.detach(), y_surf)
    loss = loss + cfg.flow_matching_weight * fm_loss
    # Log fm_loss to W&B
```

**IMPORTANT:** Use `h_surf.detach()` — the flow matching loss should NOT backpropagate through the backbone. The backbone is trained by the standard loss. The flow matching head learns from the backbone's fixed representations. This prevents competing gradient signals.

### Step 5: Inference — use flow matching samples

In the evaluation loop, after the model's standard forward pass produces `pred`:

```python
if cfg.flow_matching_surface and hasattr(model, 'flow_surface'):
    # Get backbone hidden states at surface nodes
    h_surf = fx[:, surf_idx, :]  # [B, M_s, n_hidden]
    
    # Sample N predictions and average
    fm_preds = []
    for _ in range(cfg.flow_matching_n_samples):
        sample = model.flow_surface.sample(
            h_surf.reshape(-1, h_surf.shape[-1]),  # [B*M_s, n_hidden]
            n_steps=10
        ).reshape(h_surf.shape[0], -1, 3)  # [B, M_s, 3]
        fm_preds.append(sample)
    fm_avg = torch.stack(fm_preds).mean(0)  # [B, M_s, 3]
    
    # Blend: 50% standard SRF prediction + 50% flow matching average
    pred[:, surf_idx, :] = 0.5 * pred[:, surf_idx, :] + 0.5 * fm_avg
```

### Step 6: Add argparse flags

```python
parser.add_argument('--flow_matching_surface', action='store_true')
parser.add_argument('--flow_matching_n_samples', type=int, default=8)
parser.add_argument('--flow_matching_weight', type=float, default=0.5)
parser.add_argument('--flow_matching_t_embed_dim', type=int, default=16)
```

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/flow-matching-srf-s42" \
  --wandb_group "nezuko/flow-matching-surface" \
  --seed 42 \
  --flow_matching_surface --flow_matching_weight 0.5 --flow_matching_n_samples 8 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — same with --seed 73 --wandb_name "nezuko/flow-matching-srf-s73"
```

### Report

Table: p_in, p_oodc, p_tan, p_re for both seeds and 2-seed avg vs baseline. W&B run IDs. Also report the flow matching loss value (fm_loss) at the end of training.

**Watch for:**
- If training is significantly slower (flow matching adds extra forward pass), reduce n_samples to 4 at inference
- If fm_loss doesn't decrease over training, the flow head isn't learning — try increasing flow_matching_weight to 1.0
- The key diagnostic: does the flow matching prediction VARIANCE decrease over training? If samples at inference are highly variable (std > 0.5 of prediction magnitude), the flow head hasn't converged

---

## Baseline (PR #2251, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.891 | < 11.89 |
| p_oodc | 7.561 | < 7.56 |
| p_tan | 28.118 | < 28.12 |
| p_re | 6.364 | < 6.36 |

Baseline W&B runs: `7jix2jkg` (s42), `epkfhxfl` (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Flow Matching Surface Head (AlphaFold3-inspired)

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | **11.891** | 15.2 | 14.7 | 14.95 | +3.06 ❌ |
| p_oodc | **7.561**  | 10.0 | 10.0 | 10.00 | +2.44 ❌ |
| p_tan  | **28.118** | 31.9 | 32.3 | 32.10 | +3.98 ❌ |
| p_re   | **6.364**  | 7.6  | 7.7  | 7.65  | +1.29 ❌ |

**W&B runs:**
- Seed 42: [`f70htuae`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/f70htuae)
- Seed 73: [`xo6cm5o3`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xo6cm5o3)

**Commands used:**
```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=4 python train.py --agent nezuko \
  --wandb_name "nezuko/flow-matching-srf-s42" \
  --wandb_group "nezuko/flow-matching-surface" \
  --seed 42 \
  --flow_matching_surface --flow_matching_weight 0.5 --flow_matching_n_samples 8 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 on GPU 5
```

Flow matching head: 138,371 params. Best models at epochs 141 (s42) and 143 (s73).

## Research

Researcher-agent reviewed the approach. Key findings incorporated:
- Used standard linear CFM (Lipman 2022) as recommended — OT-CFM deferred since physics regression is near-unimodal
- Used `h_surf.detach()` per AF3 practice — flow matching gradients don't backprop through backbone
- K=8 samples averaged at inference with 10 Euler steps
- 50/50 blend of SRF prediction and flow matching average

### What happened

**Strongly negative across all metrics — the worst regression we've seen.** Every metric degraded by 10-30% relative.

- **p_in degraded by +3.06** (25% regression): The 50/50 blend of SRF + flow matching predictions is catastrophic for in-distribution accuracy.
- **p_oodc degraded by +2.44** (32% regression): The hypothesis was that generative modeling would help OOD cases. Instead, it made them significantly worse.
- **p_tan degraded by +3.98** (14% regression): Tandem transfer is already the hardest metric and flow matching made it even worse.
- **p_re degraded by +1.29** (20% regression): Reynolds OOD also suffered badly.

### Why it failed

1. **The problem is nearly deterministic, not multi-modal.** The core hypothesis — that aft-foil pressure has multi-modal conditional distributions — appears to be wrong for this dataset. Given (geometry, Re, AoA, gap, stagger), the surface pressure field is essentially deterministic. Flow matching adds stochastic noise to a problem that doesn't need it, and even K=8 sample averaging doesn't fully remove it.

2. **50/50 blending corrupts the SRF prediction.** The SRF head has learned high-quality corrections after ~150 epochs. Blending it 50/50 with a generative model that produces spatially incoherent per-node samples degrades the SRF's precise corrections. The flow matching head operates per-node independently, ignoring spatial correlations on the mesh surface.

3. **Detached gradients limit flow head quality.** With `h_surf.detach()`, the flow matching head only receives fixed backbone representations. It cannot adapt the backbone to produce representations that are useful for generative sampling — only for the regression task.

4. **The auxiliary loss (weight=0.5) may destabilize early training.** Researcher-agent recommended ramping the weight from 0.1→0.5 over the first 40% of epochs. Using constant 0.5 from epoch 0 exposes early training to a high-loss flow matching signal from randomly-initialized parameters.

5. **Flow matching is designed for high-dimensional data with complex distributions** (images, protein structures). Surface pressure on a CFD mesh with ~300 surface nodes and 3 channels is low-dimensional and near-deterministic — the wrong regime for generative modeling.

### Suggested follow-ups

1. **Don't pursue flow matching for this problem.** The results are decisively negative. The conditional distribution of surface pressure is near-deterministic given the input features — generative modeling adds noise, not information.
2. **If revisited:** Use flow matching as a **post-hoc uncertainty estimator** only (don't blend into prediction), or try it only on the hardest p_tan subset with full replacement rather than blending.
3. **Focus on deterministic improvements** for p_tan: better backbone representations of wake interactions, spatially-aware SRF correction (e.g., graph-corrected refinement), or targeted hard-sample training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2271 because ADVISOR: Closing — all 4 metrics regressed significantly (p_in +25.7%, p_oodc +32.3%, p_tan +14.2%, p_re +20.2%).

**Root cause:** The core hypothesis is wrong for this problem. CFD surface pressure is near-deterministic given (geometry, Re, AoA, gap, stagger) — there is no meaningful multi-modality to model. Flow matching adds stochastic noise to a deterministic signal. The 50/50 SRF blend corrupts precise SRF corrections with spatially incoherent per-node samples. Detached gradients prevent the flow head from influencing backbone representations, further limiting its quality.

**Key takeaway:** Generative modeling (flow matching, diffusion, VAE-style) is the wrong paradigm for this problem. The conditional distribution of surface pressure is a delta function, not a distribution. Added to DO NOT REVISIT.

Excellent analysis and research integration. Moving you to a completely different direction..
